process.py: shlex module quotation escaping fix

svirt opened the pull request:
shlex.split() function fails in some particular test cases (e.g.
node_memtune.negative_testing.set_node_memory_parameter.invalid_value.
shm_sleep_millisecs) due to wrong parsing of quotation marks.
Replacing with escaped characters solves the issue.
( Corrected at the all places where shlex.split() is called )

svirt commented:
I will need to analyze while Travis fails here (looks strange) because normally there is no more ValueError after this fix is applied on my working environments. I will close this PR for now and will reopen it after the checks are done.

ldoktor commented:
Well python is actually right, the `"`, `'` and ``` are not finished. Proper fix is in my opinion to escape those chars in the tests. I had similar problems when working on: https://github.com/avocado-framework/avocado/pull/865 the problem is when we use this kind of magic in `process`, users will have problems in commads they actually need to be split (eg. `ls "my cool name"` should produce ["ls", "my cool name"]` and not `["ls", '"my', "cool", 'name"']`).

ldoktor commented:
Alternatively you can execute it as `shell=True`, which should leave it on shell.

svirt commented:
This PR can be replaces with  https://github.com/autotest/tp-libvirt/pull/669

lmr commented:
I think I'm more sympathetic to simply using `shell=True` in the `process.*` calls, since changing the configuration seems unintuitive.

svirt commented:
Probably I don't get the idea properly but in my opinion shell=True will not help as there is no dependency of that parameter at the place where the shlex fails (in process.py). And actually process.run is called already with shell=True

        # Raise exception if ignore_status is False
        ret = process.run(cmd, timeout=timeout, verbose=debug,
                          ignore_status=ignore_status,
                          shell=True)

03:59:58 stacktrace       L0042 ERROR|   File "/home/avocado-jenkins/avocado/virttest/virsh.py", line 679, in command
03:59:58 stacktrace       L0042 ERROR|     shell=True)
03:59:58 stacktrace       L0042 ERROR|   File "/home/avocado-jenkins/avocado/avocado/utils/process.py", line 948, in run
03:59:58 stacktrace       L0042 ERROR|     klass = get_sub_process_klass(cmd)
03:59:58 stacktrace       L0042 ERROR|   File "/home/avocado-jenkins/avocado/avocado/utils/process.py", line 906, in get_sub_process_klass
03:59:58 stacktrace       L0042 ERROR|     if should_run_inside_gdb(cmd):
03:59:58 stacktrace       L0042 ERROR|   File "/home/avocado-jenkins/avocado/avocado/utils/process.py", line 864, in should_run_inside_gdb
03:59:58 stacktrace       L0042 ERROR|     args = shlex.split(cmd)

One possible solution I could say is to call shles.split() with posix=False parameter

lmr commented:
@svirt it's right there in the code:

            if self.shell is False:
                cmd = shlex.split(self.cmd)
            else:
                cmd = self.cmd

lmr commented:
Basically shell=True executes the command in a subshell. That means the full string of the command can be passed to the shell.

shell=False means we're going to execute it directly, with no shell intermediary. This means the command string has to be split with shlex before passing it to the subprocess call.

lmr commented:
Now I figured out why you had problems even with shell=True -> `process.should_run_inside_gdb` also calls shlex.split to analyze whether the command should run inside gdb.

Oh well, then there's nothing we can do but escape the unclosed quotation in the config file as you did.

lmr commented:
So, to summarize - the fix you made (cfg file) is the correct way to deal with this. In the avocado side, we could create a patch to be more robust against commands with unclosed quotes (try/except on the shlex call, and in the except clause try to escape the characters and see if that works as a freebie for the test writer (and would make the avocado API more robust).

@clebergnu, see the message above. We could change `process.should_run_inside_gdb` to make it more robust against commands with unclosed quotations, since that function is going to be splitting the command regardless of the value of `shell`.

lmr commented:
So I've sent #884, although I'm not sure we should go that route myself. Just food for thought.

clebergnu commented:
It looks reasonable. I'll give it a better look in the morning. Thanks!

svirt commented:
**@lmr**
Thank you alot for the investigation. So till your PR #884 is merged I will use my local copy of modified config file, then when your PR is merged I will check my case and if it works I will close my PR  https://github.com/autotest/tp-libvirt/pull/669
